tests/ui/lint/non_local_definitions.rs

```diff
@@ -0,0 +1,246 @@
+    }
+
+    impl dyn Uto5 {}
+    //~^ WARN non local definition
+
+    impl<T: Uto5> Uto5 for Vec<T> { }
+    //~^ WARN non local definition
+
+    impl Uto5 for &dyn Uto5 {}
+    //~^ WARN non local definition
+
+    impl Uto5 for *mut Test {}
+    //~^ WARN non local definition
+
+    impl Uto5 for *mut [Test] {}
+    //~^ WARN non local definition
+
+    impl Uto5 for [Test; 8] {}
+    //~^ WARN non local definition
+
+    impl Uto5 for (Test,) {}
+    //~^ WARN non local definition
+
+    impl Uto5 for fn(Test) -> () {}
+    //~^ WARN non local definition
+
+    impl Uto5 for fn() -> Test {}
+    //~^ WARN non local definition
+
+    let _a = || {
+        impl Uto5 for Test {}
+        //~^ WARN non local definition
+
+        1
+    };
+
+    struct InsideMain;
+
+    impl Uto5 for *mut InsideMain {}
+    impl Uto5 for *mut [InsideMain] {}
+    impl Uto5 for [InsideMain; 8] {}
+    impl Uto5 for (InsideMain,) {}
+    impl Uto5 for fn(InsideMain) -> () {}
+    impl Uto5 for fn() -> InsideMain {}
+
+    impl Debug for InsideMain {
+        fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+            todo!()
+        }
+    }
+
+    impl InsideMain {
+        fn foo() {}
+    }
+
+    fn inside_inside() {
+        impl Display for InsideMain {
+        //~^ WARN non local definition
+            fn fmt(&self, _f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+                todo!()
+            }
+        }
+
+        impl InsideMain {
+        //~^ WARN non local definition
+            fn bar() {
+                #[macro_export]
+                macro_rules! m2 { () => { } };
+                //~^ WARN non local definition
+            }
+        }
+    }
+
+    trait Uto6 {}
+    impl dyn Uto6 {}
+    impl Uto5 for dyn Uto6 {}
+    impl<T: Uto6> Uto3 for Vec<T> { }
+}
+
+struct Cat;
+
+fn meow() {
+    impl From<Cat> for () {
+    //~^ WARN non local definition
+        fn from(_: Cat) -> () {
+            todo!()
+        }
+    }
+}
+
+struct Dog;
+
+fn woof() {
+    impl PartialEq<Dog> for &Dog {
+    //~^ WARN non local definition
+        fn eq(&self, _: &Dog) -> bool {
+            todo!()
+        }
+    }
+
+    impl PartialEq<()> for Dog {
+    //~^ WARN non local definition
+        fn eq(&self, _: &()) -> bool {
+            todo!()
+        }
+    }
+
+    impl PartialEq<()> for &Dog {
+    //~^ WARN non local definition
+        fn eq(&self, _: &()) -> bool {
+            todo!()
+        }
+    }
+
+    impl PartialEq<Dog> for () {
+    //~^ WARN non local definition
+        fn eq(&self, _: &Dog) -> bool {
+            todo!()
+        }
+    }
+
+    struct Test;
+    impl PartialEq<Dog> for Test {
+        fn eq(&self, _: &Dog) -> bool {
+            todo!()
+        }
+    }
+}
+
+struct Wrap<T>(T);
+
+impl Wrap<Wrap<Wrap<()>>> {}
+
+fn rawr() {
+    struct Lion;
+
+    impl From<Wrap<Wrap<Lion>>> for () {
+        fn from(_: Wrap<Wrap<Lion>>) -> Self {
+            todo!()
+        }
+    }
```

Comment:
This should lint, as per the example above, swapping `From` for `AsRef` would allow you to get `Lion` outside of `rawr`.

Reply:
Fixed.